File: algorithms/589-n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.go
package problem589

type Node struct {
	Val      int
	Children []*Node
}

// 递归
func preorder(root *Node) []int {
	var ans []int
	var dfs func(node *Node)
	dfs = func(node *Node) {
		if node == nil {
			return
		}
		ans = append(ans, node.Val)
		for _, ch := range node.Children {
			dfs(ch)
		}
	}

	dfs(root)
	return ans
}

// 迭代
func preorderV2(root *Node) []int {
	var ans []int
	stack := []*Node{}
    nextIndex := map[*Node]int{}
	node := root
	for len(stack) > 0 || node != nil {
		for node != nil {
			ans = append(ans, node.Val)
			stack = append(stack, node)
			if len(node.Children) == 0 {
				break
			}
			nextIndex[node] = 1
			node = node.Children[0]
		}
		node = stack[len(stack)-1]
		i := nextIndex[node]
		if i < len(node.Children) {
			nextIndex[node] = i + 1
			node = node.Children[i]
		} else {
			stack = stack[:len(stack)-1]
			delete(nextIndex, node)
			node = nil
		}
	}
	return ans
}

File: algorithms/589-n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.py
"""
589 n array tree preorder traversal

N 叉树的前序遍历
"""


from collections import defaultdict
from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


def preorder(root: Node) -> List[int]:
    """
    递归
    """
    ans = []

    def dfs(node: Node):
        if node is None:
            return
        ans.append(node.val)
        for ch in node.children:
            dfs(ch)
    dfs(root)
    return ans


def preorder_v2(root: Node) -> List[int]:
    """
    迭代
    """
    if root is None:
        return []
    stack = []
    ans = []
    next_index = defaultdict(int)
    node = root
    while stack or node:
        while node:
            ans.append(node.val)
            stack.append(node)
            if not node.children:
                break
            next_index[node] = 1
            node = node.children[0]
        node = stack[-1]
        i = next_index[node]
        if i < len(node.children):
            next_index[node] = i + 1
            node = node.children[i]
        else:
            stack.pop()
            del next_index[node]
            node = None
    return ans
